tests/std/tests/P2321R2_views_zip/test.cpp

```diff
@@ -0,0 +1,838 @@
+            const ZipType> == (std::ranges::sized_range<const ZipType> ||
+                std::ranges::forward_range<const ZipType>));
+        // clang-format on
+        if constexpr (CanMemberEmpty<const ZipType>) {
+            assert(std::as_const(zipped_range).empty() == is_empty);
+        }
+
+        STATIC_ASSERT(CanBool<ZipType> == CanMemberEmpty<ZipType>);
+        if constexpr (CanBool<ZipType>) {
+            assert(static_cast<bool>(zipped_range) != is_empty);
+        }
+
+        STATIC_ASSERT(CanBool<const ZipType> == CanMemberEmpty<const ZipType>);
+        if constexpr (CanBool<const ZipType>) {
+            assert(static_cast<bool>(std::as_const(zipped_range)) != is_empty);
+        }
+
+        // Validate contents of zipped range
+        assert(std::ranges::equal(zipped_range, tuple_element_arr, [](const auto& lhs_tuple, const auto& rhs_tuple) {
+            return do_tuples_reference_same_objects(lhs_tuple, rhs_tuple);
+        }));
+
+#pragma warning(push)
+#pragma warning(disable : 4127)
+        if (!(std::ranges::forward_range<AllView<RangeTypes>> && ...)) // intentionally not if constexpr
+        {
+            return true;
+        }
+#pragma warning(pop)
+
+        // Validate view_interface::data()
+        //
+        // This should never exist because zip_view's iterator never models contiguous_iterator.
+        STATIC_ASSERT(!std::contiguous_iterator<std::ranges::iterator_t<ZipType>>);
+        STATIC_ASSERT(!CanMemberData<ZipType>);
+        STATIC_ASSERT(!std::contiguous_iterator<std::ranges::iterator_t<const ZipType>>);
+        STATIC_ASSERT(!CanMemberData<const ZipType>);
+
+        // Validate view_interface::operator[]
+        STATIC_ASSERT(CanIndex<ZipType> == std::ranges::random_access_range<ZipType>);
+        if constexpr (CanIndex<ZipType>) {
+            assert(do_tuples_reference_same_objects(zipped_range[0], tuple_element_arr[0]));
+        }
+
+        STATIC_ASSERT(CanIndex<const ZipType> == std::ranges::random_access_range<const ZipType>);
+        if constexpr (CanIndex<const ZipType>) {
+            assert(do_tuples_reference_same_objects(std::as_const(zipped_range)[0], tuple_element_arr[0]));
+        }
+
+        // Validate view_interface::front()
+        STATIC_ASSERT(CanMemberFront<ZipType> == std::ranges::forward_range<ZipType>);
+        if constexpr (CanMemberFront<ZipType>) {
+            assert(do_tuples_reference_same_objects(zipped_range.front(), tuple_element_arr[0]));
+        }
+
+        STATIC_ASSERT(CanMemberFront<const ZipType> == std::ranges::forward_range<const ZipType>);
+        if constexpr (CanMemberFront<const ZipType>) {
+            assert(do_tuples_reference_same_objects(std::as_const(zipped_range).front(), const_tuple_element_arr[0]));
+        }
+
+        // Validate view_interface::back()
+        STATIC_ASSERT(
+            CanMemberBack<
+                ZipType> == (std::ranges::bidirectional_range<ZipType> && std::ranges::common_range<ZipType>) );

```

Comment:
uncomfortable clang-formatting, maybe try newline at a specific place with `//`
```suggestion
        STATIC_ASSERT(CanMemberBack<ZipType> == //
               (std::ranges::bidirectional_range<ZipType> && std::ranges::common_range<ZipType>) );
```